pkg/virt-controller/watch/ovm.go

```diff
@@ -0,0 +1,710 @@
+		vm.ObjectMeta.OwnerReferences = []v1.OwnerReference{v1.OwnerReference{
+			APIVersion:         virtv1.OfflineVirtualMachineGroupVersionKind.GroupVersion().String(),
+			Kind:               virtv1.OfflineVirtualMachineGroupVersionKind.Kind,
+			Name:               ovm.ObjectMeta.Name,
+			UID:                ovm.ObjectMeta.UID,
+			Controller:         &t,
+			BlockOwnerDeletion: &t,
+		}}
+		vm, err := c.clientset.VM(ovm.ObjectMeta.Namespace).Create(vm)
+		if err != nil {
+			log.Log.Object(ovm).Infof("Failed to create VM: %s/%s", vm.Namespace, vm.Name)
+			c.expectations.CreationObserved(ovmKey)
+			c.recorder.Eventf(ovm, k8score.EventTypeWarning, FailedCreateVirtualMachineReason, "Error creating virtual machine: %v", err)
+			return err
+		}
+		c.recorder.Eventf(ovm, k8score.EventTypeNormal, SuccessfulCreateVirtualMachineReason, "Created virtual machine: %v", vm.ObjectMeta.Name)
+	}
+
+	if ovm.Spec.Running == false {
+		log.Log.Object(ovm).Info("It is false delete")
+		if vm == nil {
+			log.Log.Info("vm is nil")
+			// vm should not run and is not running
+			return nil
+		}
+
+		// stop it
+		c.expectations.ExpectDeletions(ovmKey, []string{controller.VirtualMachineKey(vm)})
+		err := c.clientset.VM(ovm.ObjectMeta.Namespace).Delete(vm.ObjectMeta.Name, &v1.DeleteOptions{})
+		// Don't log an error if it is already deleted
+		if err != nil {
+			// We can't observe a delete if it was not accepted by the server
+			c.expectations.DeletionObserved(ovmKey, controller.VirtualMachineKey(vm))
+			c.recorder.Eventf(ovm, k8score.EventTypeWarning, FailedDeleteVirtualMachineReason, "Error deleting virtual machine %s: %v", vm.ObjectMeta.Name, err)
+			return err
+		}
+		c.recorder.Eventf(ovm, k8score.EventTypeNormal, SuccessfulDeleteVirtualMachineReason, "Deleted virtual machine: %v", vm.ObjectMeta.UID)
+		log.Log.Object(ovm).Info("Dispatching delete event")
+	}
+
+	return nil
+}
+
+// filterActiveVMs takes a list of VMs and returns all VMs which are not in a final state
+// TODO +pkotas unify with replicaset this code is the same without dependency
+func (c *OVMController) filterActiveVMs(vms []*virtv1.VirtualMachine) []*virtv1.VirtualMachine {
+	return filter(vms, func(vm *virtv1.VirtualMachine) bool {
+		return !vm.IsFinal()
+	})
+}
+
+// filterReadyVMs takes a list of VMs and returns all VMs which are in ready state.
+// TODO +pkotas unify with replicaset this code is the same
+func (c *OVMController) filterReadyVMs(vms []*virtv1.VirtualMachine) []*virtv1.VirtualMachine {
+	return filter(vms, func(vm *virtv1.VirtualMachine) bool {
+		return vm.IsReady()
+	})
+}
+
+// listVMsFromNamespace takes a namespace and returns all VMs from the VM cache which run in this namespace
+// TODO +pkotas unify this code with replicaset
+func (c *OVMController) listVMsFromNamespace(namespace string) ([]*virtv1.VirtualMachine, error) {
+	objs, err := c.vmInformer.GetIndexer().ByIndex(cache.NamespaceIndex, namespace)
```

Comment:
There's a possibly more efficient approach.
`err := cache.ListAllByNamespace(indexer, namespace, selector, func(obj interface{})`
check out presets.go for an example of how to do this.

Reply:
We should consider refactoring this at some point as there's now 3 functions within the watch module alone doing something extremely similar.

Reply:
I would address this in separate PR, where we can do refactoring for the code as a whole.
If that would be fine for you.